Add batching rule for unsqueeze, squeeze, and transpose

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40518 Disallow passing functions that don't return Tensors to vmap
* #40517 Add batching rule for Tensor.permute
* **#40455 Add batching rule for unsqueeze, squeeze, and transpose**

These don't need to be implemented right now but are useful later down
the line. I thought I would use these in implementing vmap's `out_dims`
functionality, but it turns out they weren't necessary. Since the code
exists and is useful anyways, I am leaving this PR here.

Test Plan:
- `./build/bin/vmap_test`. We could test this using the vmap frontend API,
but there is the catch that vmap cannot directly take integers right
now (all inputs passed to vmap must be Tensors at the moment). It's
possible to hack around that by declaring lambdas that take in a single
tensor argument, but those don't look nice.

Differential Revision: [D22216167](https://our.internmc.facebook.com/intern/diff/D22216167)

zou3519 committed:
Add batching rule for unsqueeze and transpose

These are necessary (I think) for implementing the vmap frontend API's
`out_dims` functionality. These are also useful aside from that because
these enable calling vmap on unsqueeze and transpose.

Test Plan:
- `./build/bin/vmap_test`. We could test this using the vmap frontend API,
but there is the catch that vmap cannot directly take integers right
now (all inputs passed to vmap must be Tensors at the moment). It's
possible to hack around that by declaring lambdas that take in a single
tensor argument, but those don't look nice.

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -98,6 +98,25 @@ Tensor expand_batching_rule(const Tensor& self, IntArrayRef size, bool implicit)
+  // self_physical.getPhysicalDim. Semantically speaking `dim` is the dimension
+  // index on the NEW tensor where the new dimension appears.
+  auto dim_physical =
+      self_physical.numBatchDims() + maybe_wrap_dim(dim, /*logical_dim*/self.dim() + 1);
```
Why is this "self.dim() + 1"?

zou3519 replied:
Unsqueeze wraps to one plus the logical dim: https://github.com/pytorch/pytorch/blob/b623bdeabb0aa8da44285d303246e7f8ac06c2a9/aten/src/ATen/native/TensorShape.cpp#L1413

I'll make that clearer in the comment and link to said code

zou3519 replied:
comment added

zou3519 committed:
Update on "Add batching rule for unsqueeze and transpose"


These are necessary (I think) for implementing the vmap frontend API's
`out_dims` functionality. These are also useful aside from that because
these enable calling vmap on unsqueeze and transpose.

Test Plan:
- `./build/bin/vmap_test`. We could test this using the vmap frontend API,
but there is the catch that vmap cannot directly take integers right
now (all inputs passed to vmap must be Tensors at the moment). It's
possible to hack around that by declaring lambdas that take in a single
tensor argument, but those don't look nice.

[ghstack-poisoned]

zou3519 committed:
Update on "Add batching rule for unsqueeze, squeeze, and transpose"


These are necessary (I think) for implementing the vmap frontend API's
`out_dims` functionality. These are also useful aside from that because
these enable calling vmap on unsqueeze and transpose.

Test Plan:
- `./build/bin/vmap_test`. We could test this using the vmap frontend API,
but there is the catch that vmap cannot directly take integers right
now (all inputs passed to vmap must be Tensors at the moment). It's
possible to hack around that by declaring lambdas that take in a single
tensor argument, but those don't look nice.

[ghstack-poisoned]

zou3519 committed:
Update on "Add batching rule for unsqueeze, squeeze, and transpose"


These don't need to be implemented right now but are useful later down
the line. I thought I would use these in implementing vmap's `out_dims`
functionality, but it turns out they weren't necessary. Since the code
exists and is useful anyways, I am leaving this PR here.

Test Plan:
- `./build/bin/vmap_test`. We could test this using the vmap frontend API,
but there is the catch that vmap cannot directly take integers right
now (all inputs passed to vmap must be Tensors at the moment). It's
possible to hack around that by declaring lambdas that take in a single
tensor argument, but those don't look nice.

[ghstack-poisoned]

zou3519 committed:
Update on "Add batching rule for unsqueeze, squeeze, and transpose"


These don't need to be implemented right now but are useful later down
the line. I thought I would use these in implementing vmap's `out_dims`
functionality, but it turns out they weren't necessary. Since the code
exists and is useful anyways, I am leaving this PR here.

Test Plan:
- `./build/bin/vmap_test`. We could test this using the vmap frontend API,
but there is the catch that vmap cannot directly take integers right
now (all inputs passed to vmap must be Tensors at the moment). It's
possible to hack around that by declaring lambdas that take in a single
tensor argument, but those don't look nice.

[ghstack-poisoned]
